### ISSUE TITLE
Graduate injection of CDI devices from experimental to GA

### ISSUE THREAD

**Issue by elezar** [moby/moby#45192]
### Description

https://github.com/moby/moby/pull/45134 and https://github.com/docker/cli/pull/4084 add experimental support for injecting CDI resources into a container. This is currently gated by the `--experimental` flag and this feature tracks what is required to promote this to GA.

*NOTE:* This issue is in "draft" and is being updated as the PRs are being reviewed.

Tasks:
- [ ] Add CDI spec dirs to `docker info`. See https://github.com/moby/moby/pull/45134#discussion_r1185363084
    - [ ] For the `moby` daemon: https://github.com/moby/moby/pull/46004
    - [ ] For the `docker` CLI: <TODO> 
- [ ] Refactor registration of device drivers to mirror what is done for the `pluginStore`. See https://github.com/moby/moby/pull/45134#discussion_r1142384094
- [ ] Improve API for requesting CDI devices in non-CLI use cases. Having to specify a driver name (`cdi`) and a capability (`cdi`), for example could be made more user friendly. See https://github.com/moby/moby/pull/45134#discussion_r1143723984
- [ ] Define a way forward for supporting Windows devices in CDI. See https://github.com/moby/moby/pull/45134#discussion_r1143647706
- [ ] Define `DefaultSpecDirs` on windows platforms. See https://github.com/container-orchestrated-devices/container-device-interface/issues/133

**Comment by gautaz** [comment]
Hello,

This [podman issue](https://github.com/containers/podman/issues/19338) seems to indicate that the CDI driver will be supported by compose.
Is there a way to track how CDI support is evolving in the docker ecosystem?

I was initially trying to find something related to CDI in https://github.com/docker/compose/issues but as nothing came out, I looked at moby's issues, so my apologies if this comment feels misplaced.

**Comment by TBBle** [comment]
@gautaz:

I don't think `docker-compose` needs to do anything to support CDI, as it's just implemented as another device driver in Docker, so it's up to the underlying Docker daemon's support, which was indeed included in the 25.0.0 release with the experimental flag.

If you try it and it doesn't work (don't forget to enable the experimental flag in your Docker daemon. ^_^), I'd suggest opening a ticket in https://github.com/docker/compose/issues so it can be determined why docker-compose did not successfully pass on the necessary flags.

As for the rest of the ecosystem, _this ticket_ is the place to track the promotion of CDI support from experimental to general availability.

**Comment by elezar** [comment]
@gautaz as @TBBle points out, CDI should be supported "out of the box" on Compose as long as the `cdi` driver is specified. I have the following (untested) example from my discussions while developing the feature:
```
services:
  test:
    image: application-image
    command: do-the-thing
    deploy:
      resources:
        reservations:
          devices:
            # NVIDIA Devices
            - driver: cdi
              deviceIDs:
                - nvidia.com/gpu=0
                - nvidia.com/gds=all
                - nvidia.com/mofed=all
            # Intel Devices
            - driver: cdi
              deviceIDs:
                - intel.com/vf=0
                - intel.com/fpga=fist
```
With regards to https://github.com/containers/podman/issues/19338 that you mentioned, this seems to be related to `podman-compose` and I have not had the time to really dig into where that source code is and what changes are required to enable CDI support there.

**Comment by gautaz** [comment]
Thanks both @TBBle and @elezar for your answers.

To clarify what I am dealing with, I am using Nixos and for now, due to [issues leading to my inability to use docker in rootless mode with a GPU](https://discourse.nixos.org/t/docker-rootless-with-nvidia-support/37069), I am in fact using podman.
But as I am also already using compose files initially developed for docker-compose, I am still using docker-compose to run them and not podman-compose for feature parity reasons.

Currently, my tests with this configuration show that:
- podman (4.9.3) is able to run containers in rootless mode with GPU access
- docker-compose (2.24.6) uses podman to run containers per my compose files
- docker-compose seems to fail to understand part of @elezar suggestion for the following compose file:

```yaml
services:
  testgpuaccess:
    image: testimage
    deploy:
      resources:
        reservations:
          devices:
            - driver: cdi
              deviceIDs:
                - nvidia.com/gpu=all
```

`podman compose` (which in fact calls `docker-compose`) fails this way:
```console
❯ podman compose up -d
>>>> Executing external compose provider "/home/user/.nix-profile/bin/docker-compose". Please refer to the documentation for details. <<<<

validating /home/user/test/kimea-saas/compose.yml: services.testgpuaccess.deploy.resources.reservations.devices.0 Additional property deviceIDs is not allowed
Error: executing /home/user/.nix-profile/bin/docker-compose up -d: exit status 15
```

I am not entirely sure where the fault lies as my configuration is an hybridation of both docker and podman solutions.

**Comment by elezar** [comment]
@gautaz please open a new issue for this since it is out of scope of this issue here.

**Comment by gautaz** [comment]
@elezar `device_ids` was in fact the right syntax, I should have looked earlier at the [compose spec](https://github.com/compose-spec/compose-spec/blob/master/deploy.md#device_ids), my apologies.

You are right, the issues I am facing now are not moby related, I'll open new issues on podman and docker-compose projects.
Thanks again for your kind answers and sorry for the pollution.

**Comment by austinvazquez** [comment]
Hey folks, I am working to come up to speed on this issue. Curious if anyone knew the current state? I see some tasks are resolved. Is Windows support the big hurdle for CDI to be GA?

**Comment by vvoland** [comment]
CDI will be enabled by default in v28.2.0.

Let me close this one and let's continue the conversation on Windows support in: https://github.com/moby/moby/issues/50045
